src/gui/guiFormSpecMenu.cpp

```diff
@@ -347,6 +348,111 @@ void GUIFormSpecMenu::parseContainerEnd(parserData* data)
 	}
 }
 
+void GUIFormSpecMenu::parseScrollContainer(parserData *data, const std::string &element)
+{
+	std::vector<std::string> parts = split(element, ';');
+
+	if (parts.size() < 4 ||
+			(parts.size() > 5 && m_formspec_version <= FORMSPEC_API_VERSION)) {
+		errorstream << "Invalid scroll_container start element (" << parts.size()
+				<< "): '" << element << "'" << std::endl;
+		return;
+	}
+
+	std::vector<std::string> v_pos  = split(parts[0], ',');
+	std::vector<std::string> v_geom = split(parts[1], ',');
+	std::string scrollbar_name = parts[2];
+	std::string direction = parts[3];
+	f32 scroll_factor = 0.1f;
+	if (parts.size() >= 5 && !parts[4].empty())
+		scroll_factor = stof(parts[4]);
+
+	MY_CHECKPOS("scroll_container", 0);
+	MY_CHECKGEOM("scroll_container", 1);
+
+	v2s32 pos = getRealCoordinateBasePos(v_pos);
+	v2s32 geom = getRealCoordinateGeometry(v_geom);
+
+	// old parent (at first: this)
+	// ^ is parent of clipper
+	// ^ is parent of mover
+	// ^ is parent of other elements
+
+	// make clipper
+	FieldSpec spec_clipper(
+		"",
+		L"",
+		L"",
+		258 + m_fields.size()
+	);
+
+	if (direction == "down")
+		scroll_factor *= -imgsize.Y;
+	else if (direction == "right")
+		scroll_factor *= -imgsize.X;
+	else
+		warningstream << "GUIFormSpecMenu::parseScrollContainer(): "
+				<< "Invalid scroll_container direction: " << direction
+				<< std::endl;
+
+	core::rect<s32> rect_clipper = core::rect<s32>(pos, pos + geom);
+
+	gui::IGUIElement *clipper = new gui::IGUIElement(EGUIET_ELEMENT, Environment,
+			data->current_parent, spec_clipper.fid, rect_clipper);
+
+	m_fields.push_back(spec_clipper);
+
+	// make mover
+	FieldSpec spec_mover(
+		"",
+		L"",
+		L"",
+		258 + m_fields.size()
+	);
+
+	core::rect<s32> rect_mover = core::rect<s32>(0, 0, geom.X, geom.Y);
+
+	GUIScrollContainer *mover = new GUIScrollContainer(Environment,
+			clipper, spec_mover.fid, rect_mover, direction, scroll_factor);
+
+	data->current_parent = mover;
+
+	m_scroll_containers.emplace_back(scrollbar_name, mover);
+
+	m_fields.push_back(spec_mover);
+
+	clipper->drop();
+
+	// remove interferring offset of normal containers
+	container_stack.push(pos_offset);
+	pos_offset.X = 0.0f;
+	pos_offset.Y = 0.0f;
+}
+
+void GUIFormSpecMenu::parseScrollContainerEnd(parserData *data)
+{
+	if (data->current_parent == this || data->current_parent->getParent() == this ||
```

Comment:
I'm not a fan of how this works, I think it would be better to make the container_stack take a struct rather than positions, which contains meta information on the container types. However, I'm fine with this given there's only one special container type currently